FGEGraphics/UISystem/UIInputLabel.cs

```diff
@@ -0,0 +1,481 @@
+        Submit
+    }
+
+    /// <summary>The text to display when the input is empty.</summary>
+    public UIElementText PlaceholderInfo;
+
+    /// <summary>The UI style of normal input content.</summary>
+    public UIElementStyle InputStyle;
+
+    /// <summary>The UI style of highlighted input content.</summary>
+    public UIElementStyle HighlightStyle;
+
+    /// <summary>Whether the input label is currently selected.</summary>
+    public bool Selected = false; // TODO: Provide a UIElement-native solution for this
+
+    /// <summary>Data internal to a <see cref="UIInputLabel"/> instance.</summary>
+    public InternalData Internal = new();
+
+    /// <summary>Fired when the text is edited by the user.</summary>
+    public Action<string> TextEdited;
+
+    /// <summary>Fired when the user submits the text content.</summary>
+    public Action<string> TextSubmitted;
+
+    /// <summary>Fired when the user de-selects the input label.</summary>
+    public Action Deselected;
+    
+    /// <summary>Gets or sets the input text content.</summary>
+    public string TextContent
+    {
+        get => Internal.TextContent;
+        set
+        {
+            Internal.SetTextContent(value);
+            UpdateText();
+        }
+    }
+
+    /// <summary>The current number of input text lines.</summary>
+    public int Lines => Internal.TextChain.Sum(piece => piece.Text.Lines.Length);
+
+    /// <summary>Data internal to a <see cref="UIInputLabel"/> instance.</summary>
+    public struct InternalData()
+    {
+        /// <summary>The raw text content.</summary>
+        public string TextContent = "";
+
+        /// <summary>The start cursor position. Acts as an anchorpoint for the end cursor.</summary>
+        public int CursorStart = 0;
+
+        /// <summary>The end cursor position.</summary>
+        public int CursorEnd = 0;
+
+        /// <summary>The drawn cursor offset location.</summary>
+        public Location CursorOffset;
+
+        /// <summary>The minimum cursor position.</summary>
+        public readonly int IndexLeft => CursorStart < CursorEnd ? CursorStart : CursorEnd;
+
+        /// <summary>The maximum cursor position.</summary>
+        public readonly int IndexRight => CursorEnd > CursorStart ? CursorEnd : CursorStart;
+
+        /// <summary>Whether a string of text is selected between the indices.</summary>
+        public readonly bool HasSelection => CursorStart != CursorEnd;
+
+        /// <summary>The text preceding <see cref="IndexLeft"/>.</summary>
+        public UIElementText TextLeft;
+
+        /// <summary>The text between <see cref="IndexLeft"/> and <see cref="IndexRight"/>.</summary>
+        public UIElementText TextBetween;
+
+        /// <summary>The text following <see cref="IndexRight"/>.</summary>
+        public UIElementText TextRight;
+
+        /// <summary>A text chain generated from <see cref="TextLeft"/>, <see cref="TextBetween"/>, and <see cref="TextRight"/>.</summary>
```

Comment:
per discord discussion this doc is apparently wrong

Reply:
Changed from `text chain of` to `text chain generated from` which is (more) accurate